sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,171 +135,189 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
+				gain_sft);
+		ret = regmap_write(rt715->mbq_regmap, reg_base + i,
+				gain_val);
 		if (ret != 0) {
 			dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
-				reg, gain_r_val);
+				reg_base + i, gain_val);
 			return ret;
 		}
 	}
 
 	return 0;
 }
 
+static int rt715_sdca_set_amp_gain_8ch_put(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	const unsigned int gain_sft = 8, boost_max = 3;
+	unsigned int reg_base = kcontrol->private_value, i;
+	unsigned int gain_val;
+	int ret;
+
+	for (i = 0; i < 8; i++) {
+		gain_val =
+			rt715_sdca_boost_gain(ucontrol->value.integer.value[i], boost_max,
+			gain_sft);
+		if (i < 7) {
+			ret = regmap_write(rt715->mbq_regmap, reg_base + i, gain_val);
+			if (ret != 0) {
+				dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+					reg_base + i, gain_val);
+				return ret;
+			}
+		} else {
+			ret = regmap_write(rt715->mbq_regmap, (reg_base - 1) | BIT(15),
+						gain_val);
+			if (ret != 0) {
+				dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+					(unsigned int)((reg_base - 1) | BIT(15)), gain_val);
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
 static int rt715_sdca_set_amp_gain_get(struct snd_kcontrol *kcontrol,
 		struct snd_ctl_elem_value *ucontrol)
 {
 	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
 	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
-	unsigned int val_l, val_r, ctl_l, ctl_r, neg_flag = 0;
+	unsigned int val, i;
 	int ret;
 
-	ret = regmap_read(rt715->mbq_regmap, mc->reg, &val_l);
-	if (ret < 0)
-		dev_err(component->dev, "Failed to read 0x%x, ret=%d\n", mc->reg, ret);
-	ret = regmap_read(rt715->mbq_regmap, mc->rreg, &val_r);
-	if (ret < 0)
-		dev_err(component->dev, "Failed to read 0x%x, ret=%d\n", mc->rreg,
-				ret);
-
-	/* L channel */
-	if (mc->shift == 8) {
-		ctl_l = (val_l >> mc->shift) / 10;
-	} else {
-		ctl_l = val_l;
-		if (ctl_l & BIT(15)) {
-			ctl_l = ~(val_l - 1) & 0xffff;
-			neg_flag = 1;
-		}
-		ctl_l *= 1000;
-		ctl_l >>= 8;
-		if (neg_flag)
-			ctl_l = mc->shift - ctl_l / RT715_SDCA_DB_STEP;
+	for (i = 0; i < 2; i++) {
+		ret = regmap_read(rt715->mbq_regmap, mc->reg + i, &val);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+				mc->reg + i, ret);
 		else
-			ctl_l = mc->shift + ctl_l / RT715_SDCA_DB_STEP;
+			ucontrol->value.integer.value[i] =
+				rt715_sdca_get_gain(val, mc->shift);
 	}
 
-	neg_flag = 0;
-	/* R channel */
-	if (mc->shift == 8) {
-		ctl_r = (val_r >> mc->shift) / 10;
-	} else {
-		ctl_r = val_r;
-		if (ctl_r & BIT(15)) {
-			ctl_r = ~(val_r - 1) & 0xffff;
-			neg_flag = 1;
-		}
-		ctl_r *= 1000;
-		ctl_r >>= 8;
-		if (neg_flag)
-			ctl_r = mc->shift - ctl_r / RT715_SDCA_DB_STEP;
+	return 0;
```

Comment:
even if all reads failed? Not sure what to do if one of the two reads fails, but if both do, returning success seems strange to me.